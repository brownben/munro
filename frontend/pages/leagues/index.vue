<script setup lang="ts">
import type { League } from '~/api-types'

const { data: leagues } = await useData<League[]>('leagues/')

useTitle({
  title: 'Leagues',
  description: 'All the leagues with results on Munro',
})
</script>

<template>
  <div>
    <Heading title="Leagues" />
    <main
      class="mx-auto max-w-screen-lg gap-8 px-6 py-10 sm:py-12 md:py-14 lg:px-8"
    >
      <div
        class="col-span-2 -mx-4 -mb-4 -mt-8 flex flex-col divide-y dark:divide-gray-600"
      >
        <League
          v-for="league in leagues"
          :key="league.name"
          :league="league"
          class="px-4 py-10"
        />
      </div>
    </main>
  </div>
</template>
