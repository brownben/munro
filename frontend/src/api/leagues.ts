import { getData, postData, putData, deleteData } from './requests'

export const getLeague = (name: string): Promise<League | null> =>
  getData<League | null>({
    apiLocation: `/api/leagues/${name}`,
    customErrorMessage: 'Problem Fetching League Details',
  })

export const getLeagues = (): Promise<League[] | null> =>
  getData<League[]>({
    apiLocation: '/api/leagues',
    customErrorMessage: 'Problem Fetching Leagues',
  })

export const createLeague = (data: LeagueForm): Promise<ServerMessage | null> =>
  postData<ServerMessage>({
    apiLocation: `/api/leagues`,
    data,
    customErrorMessage: 'Problem Creating League',
    customSuccessMessage: `League \`${data.name}\` Created`,
    customErrorHandler: true,
  })

export const updateLeague = (data: LeagueForm): Promise<ServerMessage | null> =>
  putData<ServerMessage>({
    apiLocation: `/api/leagues/${data.name}`,
    data,
    customErrorMessage: 'Problem Updating League',
    customSuccessMessage: `League \`${data.name}\` Updated`,
    customErrorHandler: true,
  })

export const deleteLeague = (name: string): Promise<ServerMessage | null> =>
  deleteData<ServerMessage>({
    apiLocation: `/api/leagues/${name}`,
    customErrorMessage: 'Problem Deleting League',
    customErrorHandler: true,
  })
