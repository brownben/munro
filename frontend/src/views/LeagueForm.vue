<!--
  League Form

  The form for Creating/ Updating Leagues

  On Create:
    - Show form

  On Edit:
    - Show Form
    - Fetch league data and display it (League Name from URL)
-->

<template>
  <div>
    <vue-headful
      v-if="create"
      title="Munro - Create League"
      description
      :head="{
        'meta': {name: 'robots', content:'noindex'},
      }"
    />
    <vue-headful
      v-else
      title="Munro - Edit League"
      description
      :head="{
        'meta': {name: 'robots', content:'noindex'},
      }"
    />
    <div v-if="!notFound">
      <h1 v-if="create">Create League</h1>
      <h1 v-if="!create">Edit League</h1>
      <form @submit.prevent="submit()">
        <label>Name:</label>
        <input v-model.trim="name" type="text">
        <label>Description:</label>
        <input v-model.trim="info" type="text">
        <label>Website:</label>
        <input v-model.trim="website" type="text">
        <label>Logo (URL):</label>
        <input v-model.trim="logo" type="text">
        <label>Coordinator:</label>
        <input v-model.trim="coordinator" type="text">
        <label>Scoring Method:</label>
        <dropdown-input
          v-model="scoringMethod"
          :list="['Position Based (100 Max)', 'Position Based (50 Max)']"
        />
        <label>Number of Events to Count:</label>
        <input
          v-model.number="numberOfCountingEvents"
          type="text"
          inputmode="numeric"
          pattern="[0-9]*"
          min="1"
        >
        <label>Courses: (Comma Separated)</label>
        <input v-model.trim="courses" type="text">
        <button v-if="create">Create League</button>
        <button v-if="!create">Update League</button>
      </form>
    </div>
    <not-found v-if="notFound" />
  </div>
</template>

<script>
import axios from 'axios'
import DropdownInput from '@/components/DropdownInput'
const NotFound = () => import('@/views/NotFound')

export default {
  components: {
    'NotFound': NotFound,
    'DropdownInput': DropdownInput,
  },

  data: function () {
    return {
      notFound: false,
      create: true,
      oldName: '',
      name: '',
      website: '',
      logo: '',
      coordinator: '',
      scoringMethod: '',
      numberOfCountingEvents: 1,
      courses: '',
      info: '',
      numberOfEvents: 0,
    }
  },

  // On Load
  mounted: function () {
    if (this.$route.path.includes('edit')) {
      this.create = false
      this.getLeagueDetails()
    }
  },

  methods: {
    validateForm: function () {
      return (this.name !== '' && this.scoringMethod !== '')
    },

    returnToLeaguePage: function (response) {
      this.$messages.addMessage(response.data.message)
      this.$router.push('/leagues/' + this.name)
    },

    // Scoring Method is stored in a shorthand form in database, transfer between these two forms so it is displayed in a clear manner
    scoringMethodShorthandToFull: value => {
      if (value === 'position') return 'Position Based (100 Max)'
      else if (value === 'position50') return 'Position Based (50 Max)'
      else if (value === 'position99') return 'Position Based (99 Max)'
      else if (value === 'position99average') return 'Position Based (99 Max, Reduced in a Draw)'
      else return ''
    },

    scoringMethodFullToShorthand: value => {
      if (value === 'Position Based (100 Max)') return 'position'
      else if (value === 'Position Based (50 Max)') return 'position50'
      else if (value === 'Position Based (99 Max)') return 'position99'
      else if (value === 'Position Based (99 Max, Reduced in a Draw)') return 'position99average'
      return ''
    },

    getLeagueDetails: function () {
      return axios.get('/api/leagues/' + this.$route.params.name)
        .then(response => {
          if (!response.data) this.notFound = true
          else {
            this.oldName = response.data.name
            this.name = response.data.name
            this.website = response.data.website
            this.coordinator = response.data.coordinator
            this.scoringMethod = this.scoringMethodShorthandToFull(response.data.scoringMethod)
            this.numberOfCountingEvents = response.data.numberOfCountingEvents
            this.logo = response.data.logo
            if (response.data.courses) this.courses = response.data.courses.join(',')
            this.info = response.data.description
          }
        })
        .catch(() => this.$messages.addMessage('Problem Fetching League Details'))
    },

    submit: function () {
      if (this.create) this.createLeague()
      else this.updateLeague()
    },

    createLeague: function () {
      if (this.validateForm()) {
        return axios.post('/api/leagues', {
          name: this.name,
          website: this.website,
          logo: this.logo,
          coordinator: this.coordinator,
          scoringMethod: this.scoringMethodFullToShorthand(this.scoringMethod),
          numberOfCountingEvents: this.numberOfCountingEvents,
          courses: this.courses,
          description: this.info,
        })
          .then(response => this.returnToLeaguePage(response))
          .catch(() => this.$messages.addMessage('Error: Problem Creating League - Please Try Again'))
      }
      else this.$messages.addMessage('Please Ensure Name and Scoring Method Fields are not Blank')
    },

    updateLeague: function () {
      if (this.validateForm()) {
        return axios.put('/api/leagues/' + this.oldName, {
          oldName: this.oldName,
          name: this.name,
          website: this.website,
          logo: this.logo,
          coordinator: this.coordinator,
          scoringMethod: this.scoringMethodFullToShorthand(this.scoringMethod),
          numberOfCountingEvents: this.numberOfCountingEvents,
          courses: this.courses,
          description: this.info,
        })
          .then(response => this.returnToLeaguePage(response))
          .catch(() => this.$messages.addMessage('Error: Problem Updating League - Please Try Again'))
      }
      else this.$messages.addMessage('Please Ensure Name and Scoring Method Fields are not Blank')
    },
  },
}

</script>
<style lang="stylus" scoped>
@import '../assets/styles/helpers.styl'
@import '../assets/styles/inputs.styl'

h1
  margin-bottom: 0.5rem

#error
  margin: 0.5rem 0
  padding: 0.2rem 0
  border: 1px solid purple-400
  color: purple-400
  text-align: center

.fade-enter-active, .fade-leave-active
  transition: opacity 0.5s

.fade-enter, .fade-leave-to
  opacity: 0
</style>
