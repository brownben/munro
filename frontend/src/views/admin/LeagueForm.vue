<template>
  <Layout
    :title="title"
    :not-found="!loading && !league?.name && $route.path.includes('/edit')"
  >
    <Meta :title="`Munro - ${title}`" description="" :block-robots="true" />
    <form class="col-span-2" @submit.prevent="submit">
      <InputText
        v-model.trim="league.name"
        label="Name:"
        :validators="[RequiredField('a name'), IsValidURLParameter('a name')]"
      />
      <InputNumber
        v-model.number="league.year"
        label="Year:"
        :max="2050"
        class="mt-4"
      />
      <InputText
        v-model.trim="league.description"
        label="Description:"
        class="mt-4"
      />
      <InputText
        v-model.trim="league.website"
        label="Website: (URL)"
        type="url"
        class="mt-4"
      />
      <InputText
        v-model.trim="league.coordinator"
        label="Coordinator:"
        class="mt-4"
      />
      <InputDropdown
        v-model="league.scoringMethod"
        :listWithDifferentValue="scoringMethodOptions"
        :include-blank="false"
        label="Scoring Method:"
        class="mt-4"
        :validator="RequiredField('a Scoring Method', true)"
      />
      <InputDropdown
        v-model="league.leagueScoring"
        :listWithDifferentValue="[
          { value: 'course', text: 'Per Course' },
          { value: 'overall', text: 'Overall' },
          { value: 'ageClass', text: 'Age Class' },
        ]"
        :include-blank="false"
        label="League Results:"
        class="mt-4"
      />
      <InputText
        v-model.trim="league.clubRestriction"
        label="Only Include Results from Club:"
        class="mt-4"
      />
      <InputNumber
        v-model.number="league.numberOfCountingEvents"
        :min="1"
        label="Number of Events to Count:"
        class="mt-4"
      />
      <InputText
        v-model.trim="league.courses"
        label="Courses: (Comma Separated)"
        class="mt-4"
      />
      <InputDropdown
        v-model="league.subLeagueOf"
        :list="leaguesSuitableForSubLeague"
        :include-blank="true"
        label="Sub League Of:"
        class="mt-4"
      />
      <InputTextarea
        v-model.trim="league.moreInformation"
        label="More Information:"
        class="mt-4"
      />
      <InputTextarea
        v-model="league.additionalSettings"
        label="Additional Settings: (Advanced Use Only)"
        class="mt-6 text-left"
        :validators="[IsValidJSON('additional settings')]"
      />
      <button v-if="$route.path.includes('/edit')" class="mt-8 button-lg">
        Update League
      </button>
      <button v-else class="mt-8 button-lg">Create League</button>
    </form>
  </Layout>
</template>

<script lang="ts" setup>
import { ref, computed, watchEffect } from 'vue'
import { useRoute, useRouter } from 'vue-router'

import Layout from '../../components/Layout.vue'
import InputDropdown from '../../components/InputDropdown.vue'
import InputText from '../../components/InputText.vue'
import InputTextarea from '../../components/InputTextarea.vue'
import InputNumber from '../../components/InputNumber.vue'

import { toSingleString } from '../../scripts/typeHelpers'
import {
  RequiredField,
  IsValidURLParameter,
  IsValidJSON,
} from '../../scripts/inputValidation'

import { useAuthentication } from '../../store/authentication'
import { useMessages } from '../../store/messages'

import {
  useLeagues,
  createLeague as apiCreateLeague,
  updateLeague as apiUpdateLeague,
  useLeague,
} from '../../api/leagues'

const router = useRouter()
const route = useRoute()
const messages = useMessages()
const auth = useAuthentication()

const routeParamsName = computed(() => toSingleString(route.params.name))
const [leagues] = useLeagues()
const [leagueRaw, loading] = useLeague(routeParamsName)
const league = ref<LeagueForm>({
  courses: '',
  coordinator: '',
  description: '',
  dynamicEventResults: true,
  moreInformation: '',
  name: '',
  numberOfCountingEvents: 1,
  scoringMethod: '',
  website: '',
  year: 2000,
  leagueScoring: 'course',
  clubRestriction: '',
  subLeagueOf: '',
  additionalSettings: '',
})

watchEffect(() => {
  if (leagueRaw.value)
    league.value = {
      ...leagueRaw.value,
      courses: leagueRaw.value.courses?.join(',') ?? '',
      moreInformation:
        leagueRaw.value.moreInformation?.replace(/\|\s*/g, '\n') ?? '',
      dynamicEventResults: leagueRaw.value.dynamicEventResults ?? true,
      leagueScoring: leagueRaw.value.leagueScoring ?? 'course',
    } as LeagueForm
})

const validateForm = () => {
  if (league.value.name === '' || league.value.scoringMethod === '') {
    messages.create(
      'Please Ensure Name and Scoring Method Fields are not Blank'
    )
    return false
  } else if (
    league.value.name.includes('/') ||
    league.value.name.includes('\\')
  ) {
    messages.create("Please Ensure Name doesn't Include any Slashes")
    return false
  } else return true
}
const createLeague = () => {
  if (validateForm())
    return apiCreateLeague({
      ...league.value,
      courses: league.value?.courses?.split(',') ?? [],
      moreInformation: league.value?.moreInformation?.replace(/\n/g, '|') ?? '',
    })
      .then(() => router.push(`/leagues/${league.value.name}`))
      .catch(() => false)
}
const updateLeague = () => {
  if (validateForm())
    return apiUpdateLeague({
      ...league.value,
      courses: league.value?.courses?.split(',') ?? [],
      moreInformation: league.value?.moreInformation?.replace(/\n/g, '|') ?? '',
    })
      .then(() => router.push(`/leagues/${league.value.name}`))
      .catch(() => false)
}
const submit = () =>
  route.path.includes('/edit') ? updateLeague() : createLeague()

const title = computed(() =>
  route.path.includes('/edit') ? 'Edit League' : 'Create League'
)

const leaguesSuitableForSubLeague = computed(() =>
  leagues.value
    .filter((l) => l.scoringMethod === league.value.scoringMethod)
    .map((league) => league.name)
)

const scoringMethodOptions = [
  { value: 'position', text: 'Position Based (100 Max)' },
  { value: 'position50', text: 'Position Based (50 Max)' },
  { value: 'position99', text: 'Position Based (99 Max)' },
  {
    value: 'position99average',
    text: 'Position Based (99 Max, Reduced in a Draw)',
  },
  {
    value: 'positionDouble',
    text: 'Position Based (100 Max, Double Points',
  },
  {
    value: 'position50Double',
    text: 'Position Based (50 Max, Double Points)',
  },
  {
    value: 'positionStaggered',
    text: 'Position Based (Staggered, 60 Max)',
  },
  {
    value: 'timeAverage',
    text: 'Relative to Average Time (1000 Average)',
  },
  {
    value: 'timeAverage100',
    text: 'Relative to Average Time (100 Average)',
  },
  { value: 'timeTop3', text: 'Relative to Top 3 Times' },
  {
    value: 'timeTop3Adjusted',
    text: 'Relative to Top 3 Times (Adjusted)',
  },
  { value: 'file', text: 'From Upload File' },
]
</script>
