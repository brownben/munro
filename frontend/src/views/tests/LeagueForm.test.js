/*
  League Form View Unit Tests
*/

import axios from 'axios'
import { mount } from '@vue/test-utils'
import LeagueForm from '@/views/LeagueForm'
import { sampleSingleLeague } from '@/tests/test data/leagues'

jest.mock('axios')

beforeEach(() => {
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: [] })
})

test('Is a Vue Instance', () => {
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '', params: { name: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Creation Form - Renders Correctly', () => {
  const mockGetLeagueDetails = jest.fn()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '', params: { name: '' } },
      $messages: { addMessage: jest.fn() },
    },
    methods: {
      getLeagueDetails: mockGetLeagueDetails,
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  expect(wrapper.vm.create).toBeTruthy()
  expect(mockGetLeagueDetails).toHaveBeenCalledTimes(0)
  expect(wrapper.element).toMatchSnapshot()
})

test('Update Form - Renders Correctly', () => {
  const mockGetLeagueDetails = jest.fn()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/leagues/Test/edit', params: { name: '' } },
      $messages: { addMessage: jest.fn() },
    },
    methods: {
      getLeagueDetails: mockGetLeagueDetails,
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  expect(wrapper.vm.create).toBeFalsy()
  expect(mockGetLeagueDetails).toHaveBeenCalledTimes(1)
  expect(wrapper.element).toMatchSnapshot()
})

test('Submit Method', () => {
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '', params: { name: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  const getCreateSpy = jest.spyOn(wrapper.vm, 'createLeague')
  const getUpdateSpy = jest.spyOn(wrapper.vm, 'updateLeague')

  wrapper.setData({ create: true })
  wrapper.vm.submit()
  expect(getCreateSpy).toHaveBeenCalledTimes(1)
  expect(getUpdateSpy).toHaveBeenCalledTimes(0)

  jest.clearAllMocks()

  wrapper.setData({ create: false })
  wrapper.vm.submit()
  expect(getUpdateSpy).toHaveBeenCalledTimes(1)
  expect(getCreateSpy).toHaveBeenCalledTimes(0)
})

test('Check Validation Works', () => {
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '', params: { name: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  wrapper.setData({ name: '', scoringMethod: '' })
  expect(wrapper.vm.validateForm()).toBeFalsy()

  wrapper.setData({ name: 'Value', scoringMethod: '' })
  expect(wrapper.vm.validateForm()).toBeFalsy()

  wrapper.setData({ name: '', scoringMethod: 'Value' })
  expect(wrapper.vm.validateForm()).toBeFalsy()

  wrapper.setData({ name: 'A Value', scoringMethod: 'Another Value' })
  expect(wrapper.vm.validateForm()).toBeTruthy()
})

test('Return to League Home Page', () => {
  const mockAddMessageFunction = jest.fn()
  const mockRouterPushFunction = jest.fn()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $router: { push: mockRouterPushFunction },
      $route: { path: '', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  wrapper.setData({ name: 'Test League' })
  wrapper.vm.returnToLeaguePage({ data: { message: 'This is the Message Returned' } })
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith('This is the Message Returned')
  expect(mockRouterPushFunction).toHaveBeenCalledTimes(1)
  expect(mockRouterPushFunction).toHaveBeenLastCalledWith('/leagues/Test League')
})

test('Scoring Method Shorthand to Full', () => {
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '', params: { name: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  expect(wrapper.vm.scoringMethodShorthandToFull('position')).toBe('Position Based (100 Max)')
  expect(wrapper.vm.scoringMethodShorthandToFull('position50')).toBe('Position Based (50 Max)')
  expect(wrapper.vm.scoringMethodShorthandToFull('')).toBe('')
  expect(wrapper.vm.scoringMethodShorthandToFull('pos')).toBe('')
})

test('Scoring Method Full to Shorthand', () => {
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '', params: { name: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  expect(wrapper.vm.scoringMethodFullToShorthand('Position Based (100 Max)')).toBe('position')
  expect(wrapper.vm.scoringMethodFullToShorthand('Position Based (50 Max)')).toBe('position50')
  expect(wrapper.vm.scoringMethodFullToShorthand('')).toBe('')
  expect(wrapper.vm.scoringMethodFullToShorthand('pos')).toBe('')
})

test('Create League - Fails Validation', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/create-league', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  await wrapper.vm.createLeague()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith('Please Ensure Name and Scoring Method Fields are not Blank')
})

test('Create League - Successful Creation', async () => {
  const mockAddMessageFunction = jest.fn()
  const mockReturnToLeaguePage = jest.fn()
  axios.post.mockResolvedValue({ data: { message: 'Hello' } })
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/create-league', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  wrapper.setData({ name: 'Test League', scoringMethod: 'position' })
  wrapper.setMethods({ returnToLeaguePage: mockReturnToLeaguePage })
  await wrapper.vm.createLeague()
  expect(mockReturnToLeaguePage).toHaveBeenCalledTimes(1)
  expect(mockReturnToLeaguePage).toHaveBeenLastCalledWith({ data: { message: 'Hello' } })
})

test('Create League - Error in Creation', async () => {
  const mockAddMessageFunction = jest.fn()
  axios.post.mockRejectedValue()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/create-league', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  wrapper.setData({ name: 'Test League', scoringMethod: 'position' })
  await wrapper.vm.createLeague()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith('Error: Problem Creating League - Please Try Again')
})

test('Create League - Calls Correct API Location', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/create-league', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  jest.clearAllMocks()
  axios.post.mockResolvedValue()
  const sampleOutput = Object.assign({}, sampleSingleLeague[0])
  delete sampleOutput.info
  sampleOutput.scoringMethod = 'position'
  delete sampleOutput.numberOfEvents
  delete sampleOutput.oldName
  const sampleInput = Object.assign({}, sampleSingleLeague[0])
  delete sampleInput.oldName
  sampleInput.info = sampleInput.description
  delete sampleInput.description
  sampleInput.scoringMethod = 'Position Based (100 Max)'
  wrapper.setData(sampleInput)
  await wrapper.vm.createLeague()
  expect(axios.post).toHaveBeenCalledTimes(1)
  expect(axios.post).toHaveBeenLastCalledWith('/api/leagues', sampleOutput)
})

test('Update League - Calls Correct API Location', async () => {
  const mockAddMessageFunction = jest.fn()
  axios.put.mockResolvedValue()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/leagues/1/edit', params: { name: 'name' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  const sampleOutput = Object.assign({}, sampleSingleLeague[0])
  delete sampleOutput.info
  sampleOutput.scoringMethod = 'position'
  sampleOutput.oldName = 'name'
  delete sampleOutput.numberOfEvents
  const sampleInput = Object.assign({}, sampleSingleLeague[0])
  sampleInput.info = sampleInput.description
  delete sampleInput.description
  sampleInput.oldName = 'name'
  sampleInput.scoringMethod = 'Position Based (100 Max)'
  wrapper.setData(sampleInput)
  await wrapper.vm.updateLeague()
  expect(axios.put).toHaveBeenCalledTimes(1)
  expect(axios.put).toHaveBeenLastCalledWith('/api/leagues/name', sampleOutput)
})

test('Update League - Fails Validation', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/leagues/1/edit', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  await wrapper.vm.updateLeague()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith('Please Ensure Name and Scoring Method Fields are not Blank')
})

test('Update League - Successful Creation', async () => {
  const mockAddMessageFunction = jest.fn()
  const mockGetLeagueDetails = jest.fn()
  axios.put.mockResolvedValue({ data: { message: 'Hello' } })
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/leagues/1/edit', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
      getLeagueDetails: mockGetLeagueDetails,
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  const mockReturnToLeaguePage = jest.fn()
  wrapper.setMethods({ returnToLeaguePage: mockReturnToLeaguePage })
  const sampleInput = Object.assign({}, sampleSingleLeague[0])
  sampleInput.info = sampleInput.description
  delete sampleInput.description
  wrapper.setData(sampleInput)
  await wrapper.vm.updateLeague()
  expect(mockReturnToLeaguePage).toHaveBeenCalledTimes(1)
  expect(mockReturnToLeaguePage).toHaveBeenLastCalledWith({ data: { message: 'Hello' } })
})

test('Update League - Error in Creation', async () => {
  const mockAddMessageFunction = jest.fn()
  const mockGetLeagueDetails = jest.fn()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/leagues/1/edit', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
      getLeagueDetails: mockGetLeagueDetails,
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  wrapper.setData({ name: 'Test League', scoringMethod: 'position' })
  axios.put.mockRejectedValue()
  await wrapper.vm.updateLeague()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith('Error: Problem Updating League - Please Try Again')
})

test('Get League Details - Error', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/leagues/1/edit', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
      created: () => { },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  jest.clearAllMocks()
  axios.get.mockRejectedValue()
  await wrapper.vm.getLeagueDetails()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith('Problem Fetching League Details')
})

test('Get League Details - Success', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/leagues/1/edit', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
      created: () => { },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: { ...sampleSingleLeague[0], description: '' } })
  await wrapper.vm.getLeagueDetails()
  expect(wrapper.vm.notFound).toBeFalsy()
  expect(wrapper.vm.logo).toBe(sampleSingleLeague[0].logo)
  expect(wrapper.vm.info).toBe(sampleSingleLeague[0].description)
  expect(wrapper.vm.website).toBe(sampleSingleLeague[0].website)
  expect(wrapper.vm.courses).toBe(sampleSingleLeague[0].courses.join(','))
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(0)
})

test('Get League Details - Not Found', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/leagues/1/edit', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
      created: () => { },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: false })
  await wrapper.vm.getLeagueDetails()
  expect(wrapper.vm.notFound).toBeTruthy()
  expect(wrapper.findAll('form').length).toBe(0)
})

test('Get League Details - Correct API Call', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(LeagueForm, {
    mocks: {
      $route: { path: '/leagues/1/edit', params: { name: 'a' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  jest.clearAllMocks()
  axios.get.mockRejectedValue()
  await wrapper.vm.getLeagueDetails()
  expect(axios.get).toHaveBeenCalledTimes(1)
  expect(axios.get).toHaveBeenLastCalledWith('/api/leagues/a')
})
