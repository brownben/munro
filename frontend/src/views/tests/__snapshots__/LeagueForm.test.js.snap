// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creation Form - Renders Correctly 1`] = `
<div>
  <vue-headful-stub
    description=""
    head="[object Object]"
    title="Munro - Create League"
  />
   
  <div>
    <h1>
      Create League
    </h1>
     
    <!---->
     
    <form>
      <label>
        Name:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Description:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Website:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Logo (URL):
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Coordinator:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Scoring Method:
      </label>
       
      <dropdown-input-stub
        label=""
        list="Position Based (100 Max),Position Based (50 Max),Position Based (99 Max),Position Based (99 Max - Reduced in a Draw)"
        value=""
      />
       
      <label>
        Number of Events to Count:
      </label>
       
      <input
        inputmode="numeric"
        min="1"
        pattern="[0-9]*"
        type="text"
      />
       
      <label>
        Courses: (Comma Separated)
      </label>
       
      <input
        type="text"
      />
       
      <button>
        Create League
      </button>
       
      <!---->
    </form>
  </div>
   
  <!---->
</div>
`;

exports[`Update Form - Renders Correctly 1`] = `
<div>
  <vue-headful-stub
    description=""
    head="[object Object]"
    title="Munro - Edit League"
  />
   
  <div>
    <!---->
     
    <h1>
      Edit League
    </h1>
     
    <form>
      <label>
        Name:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Description:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Website:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Logo (URL):
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Coordinator:
      </label>
       
      <input
        type="text"
      />
       
      <label>
        Scoring Method:
      </label>
       
      <dropdown-input-stub
        label=""
        list="Position Based (100 Max),Position Based (50 Max),Position Based (99 Max),Position Based (99 Max - Reduced in a Draw)"
        value=""
      />
       
      <label>
        Number of Events to Count:
      </label>
       
      <input
        inputmode="numeric"
        min="1"
        pattern="[0-9]*"
        type="text"
      />
       
      <label>
        Courses: (Comma Separated)
      </label>
       
      <input
        type="text"
      />
       
      <!---->
       
      <button>
        Update League
      </button>
    </form>
  </div>
   
  <!---->
</div>
`;
